Apromore-ProcessMining-Collection/prom-bpmn/src/main/java/org/apromore/processmining/models/graphbased/directed/bpmn/BPMNDiagramImpl.java

```diff
@@ -280,6 +281,52 @@ protected Map<DirectedGraphElement, DirectedGraphElement> cloneFrom(
 		return mapping;
 	}
 
+	@Override
+	public Map<BPMNNode, BPMNNode> cloneSubProcessContents(SubProcess subProcess) {
+		BPMNDiagram bpmnDiagram = (BPMNDiagram) subProcess.getGraph();
+		Map<BPMNNode, BPMNNode> nodeMap = new HashMap<>();
+
+		//Nodes
+		for (ContainableDirectedGraphElement subProcessChild : subProcess.getChildren()) {
+			if (subProcessChild instanceof BPMNNode) {
+				BPMNNode node = (BPMNNode) subProcessChild;
+				nodeMap.put(node, node.addCopyToDiagram(this));
+			}
```

Comment:
As discussed, this code is actually: first we need node to return a copy BPMNNode from itself called copyNode, then the "this" BPMN diagram will add this copyNode to itself. The code can be outlined below.
BPMN copyNode = node.copy().
this.addNode(copyNode).
BPMNNode.copy() and BPMNDiagram.addNode() need to be implemented.
Another issue is in the BPMNDiagram.addNode method, atm, it has to do if (node instanceof XXX) where XXX is all types of nodes. We should be looking for an improvement for this if...else block.